### ISSUE TITLE
crash on Windows/MSYS2

### ISSUE THREAD

**Issue by novomesk** [libjxl/libjxl#165]
Hello,

djxl crashes on some (not all) JXL files during decoding.

```
$ djxl jpegxl-logo.jxl output.png --num_threads=1
Segmentation fault
```

Test file: [jpegxl-logo.jxl](https://github.com/novomesk/qt-jpegxl-image-plugin/blob/main/testfiles/jpegxl-logo.jxl)

```
(gdb) run
Starting program: C:\msys64\mingw64\bin\djxl.exe jpegxl-logo.jxl output.png "--num_threads=1"
[New Thread 10912.0x3630]
[New Thread 10912.0x32d0]
[New Thread 10912.0x1a7c]
[New Thread 10912.0xc94]

Thread 5 received signal SIGSEGV, Segmentation fault.
[Switching to Thread 10912.0xc94]
0x00007ff64ad95967 in jxl::N_AVX2::ComputePixelChannel<hwy::N_AVX2::Simd<float, 8ull> >(hwy::N_AVX2::Simd<float, 8ull>, float, float const*, float const*, float const*, decltype (Zero((hwy::N_AVX2::Simd<float, 8ull>)()))*, decltype (Zero((hwy::N_AVX2::Simd<float, 8ull>)()))*, decltype (Zero((hwy::N_AVX2::Simd<float, 8ull>)()))*, unsigned long long) (x=8, gap=0xfddbdfb7e0, sm=0xfddbdfb840,
    mc=0xfddbdfb8a0, row_bottom=0x2414425fb80, row=0x2414425f980,
    row_top=0x2414425f780, dc_factor=0.000206910816, d=...)
    at C:/msys64/home/daniel/libjxl/lib/jxl/compressed_dc.cc:96
96        *gap = MaxWorkaround(*gap, Abs((*mc - *sm) / dc_quant));
(gdb) bt
#0  0x00007ff64ad95967 in jxl::N_AVX2::ComputePixelChannel<hwy::N_AVX2::Simd<float, 8ull> >(hwy::N_AVX2::Simd<float, 8ull>, float, float const*, float const*, float const*, decltype (Zero((hwy::N_AVX2::Simd<float, 8ull>)()))*, decltype (Zero((hwy::N_AVX2::Simd<float, 8ull>)()))*, decltype (Zero((hwy::N_AVX2::Simd<float, 8ull>)()))*, unsigned long long) (x=8, gap=0xfddbdfb7e0,
    sm=0xfddbdfb840, mc=0xfddbdfb8a0, row_bottom=0x2414425fb80,
    row=0x2414425f980, row_top=0x2414425f780, dc_factor=0.000206910816, d=...)
    at C:/msys64/home/daniel/libjxl/lib/jxl/compressed_dc.cc:96
#1  jxl::N_AVX2::ComputePixel<hwy::N_AVX2::Simd<float, 8ull> > (x=8,
    out_rows=0xfddbdfd310, rows_bottom=0xfddbdfd330, rows=0xfddbdfd350,
    rows_top=0xfddbdfd370, dc_factors=0xfddb5fd400)
    at C:/msys64/home/daniel/libjxl/lib/jxl/compressed_dc.cc:114
#2  operator() (__closure=0xfddb5fc200, y=1)
    at C:/msys64/home/daniel/libjxl/lib/jxl/compressed_dc.cc:188
#3  0x00007ff64adb4090 in jxl::ThreadPool::RunCallState<jxl::Status(long long unsigned int), jxl::N_AVX2::AdaptiveDCSmoothing(float const*, jxl::Image3F*, jxl::ThreadPool*)::<lambda(int, int)> >::CallDataFunc(void *, uint32_t, size_t) (
    jpegxl_opaque=0xfddb5fc0b0, value=1, thread_id=0)
    at C:/msys64/home/daniel/libjxl/lib/jxl/base/data_parallel.h:88
#4  0x00007ff64ae06bb2 in jpegxl::ThreadParallelRunner::RunRange (
    self=0xfddb5ff930, command=4294967439, thread=0)
    at C:/msys64/home/daniel/libjxl/lib/threads/thread_parallel_runner_internal.cc:137
#5  0x00007ff64ae06cb8 in jpegxl::ThreadParallelRunner::ThreadFunc (
    self=0xfddb5ff930, thread=0)
    at C:/msys64/home/daniel/libjxl/lib/threads/thread_parallel_runner_internal.cc:167
#6  0x00007ff64b4977aa in std::__invoke_impl<void, void (*)(jpegxl::ThreadParallelRunner*, int), jpegxl::ThreadParallelRunner*, unsigned int> (
    __f=@0x241441e31d8: 0x7ff64ae06bc0 <jpegxl::ThreadParallelRunner::ThreadFunc(jpegxl::ThreadParallelRunner*, int)>)
    at C:/msys64/mingw64/include/c++/10.3.0/bits/invoke.h:60
#7  0x00007ff64b4ae2c0 in std::__invoke<void (*)(jpegxl::ThreadParallelRunner*, int), jpegxl::ThreadParallelRunner*, unsigned int> (
    __fn=@0x241441e31d8: 0x7ff64ae06bc0 <jpegxl::ThreadParallelRunner::ThreadFunc(jpegxl::ThreadParallelRunner*, int)>)
    at C:/msys64/mingw64/include/c++/10.3.0/bits/invoke.h:95
#8  0x00007ff64b45bbc0 in std::thread::_Invoker<std::tuple<void (*)(jpegxl::ThreadParallelRunner*, int), jpegxl::ThreadParallelRunner*, unsigned int> >::_M_invoke<0ull, 1ull, 2ull> (this=0x241441e31c8)
    at C:/msys64/mingw64/include/c++/10.3.0/thread:264
#9  0x00007ff64b45bbe7 in std::thread::_Invoker<std::tuple<void (*)(jpegxl::ThreadParallelRunner*, int), jpegxl::ThreadParallelRunner*, unsigned int> >::operator() (this=0x241441e31c8) at C:/msys64/mingw64/include/c++/10.3.0/thread:271
#10 0x00007ff64b45b94c in std::thread::_State_impl<std::thread::_Invoker<std::tuple<void (*)(jpegxl::ThreadParallelRunner*, int), jpegxl::ThreadParallelRunner*, unsigned int> > >::_M_run (this=0x241441e31c0)
    at C:/msys64/mingw64/include/c++/10.3.0/thread:215
#11 0x00007ff81eed06e1 in ?? () from C:\msys64\mingw64\bin\libstdc++-6.dll
#12 0x00007ff821d54f33 in ?? () from C:\msys64\mingw64\bin\libwinpthread-1.dll
#13 0x00007ff8340daf5a in msvcrt!_beginthreadex ()
   from C:\WINDOWS\System32\msvcrt.dll
#14 0x00007ff8340db02c in msvcrt!_endthreadex ()
   from C:\WINDOWS\System32\msvcrt.dll
#15 0x00007ff833827034 in KERNEL32!BaseThreadInitThunk ()
   from C:\WINDOWS\System32\kernel32.dll
#16 0x00007ff834362651 in ntdll!RtlUserThreadStart ()
   from C:\WINDOWS\SYSTEM32\ntdll.dll
#17 0x0000000000000000 in ?? ()
```

configuration:

`cmake -G "MSYS Makefiles" -DCMAKE_INSTALL_PREFIX=/mingw64 -DCMAKE_BUILD_TYPE=Debug -DJPEGXL_ENABLE_PLUGINS=OFF -DBUILD_TESTING=OFF -DJPEGXL_WARNINGS_AS_ERRORS=OFF -DJPEGXL_ENABLE_SJPEG=OFF -DJPEGXL_ENABLE_BENCHMARK=OFF -DJPEGXL_ENABLE_EXAMPLES=OFF -DJPEGXL_ENABLE_MANPAGES=OFF -DJPEGXL_FORCE_SYSTEM_BROTLI=ON ..`

I observed that when I added `-DHWY_COMPILE_ONLY_SCALAR` to CXX_FLAGS, the djxl worked correctly.

**Comment by eustas** [comment]
Will take a look. Thanks for the heads-up.

**Comment by eustas** [comment]
Was able to reproduce. Will investigate later.

**Comment by novomesk** [comment]
Another backtrace on different file:
```
(gdb) run
Starting program: C:\msys64\mingw64\bin\djxl.exe sample1.jxl output.png "--num_threads=0"
[New Thread 1728.0x31f8]
[New Thread 1728.0x2350]
[New Thread 1728.0x1e1c]

Thread 1 received signal SIGSEGV, Segmentation fault.
0x00007ff7dd6c2f5b in jxl::N_AVX2::DecodeGroupImpl (get_block=0x61f4ffa830,
    group_dec_cache=0x1be61909680, dec_state=0x61f4ffd3a0, thread=0,
    group_idx=0, decoded=0x1be61905a20, draw=jxl::kDraw)
    at C:/msys64/home/daniel/libjxl/lib/jxl/dec_group.cc:413
413               dequant_block(
(gdb) bt
#0  0x00007ff7dd6c2f5b in jxl::N_AVX2::DecodeGroupImpl (
    get_block=0x61f4ffa830, group_dec_cache=0x1be61909680,
    dec_state=0x61f4ffd3a0, thread=0, group_idx=0, decoded=0x1be61905a20,
    draw=jxl::kDraw) at C:/msys64/home/daniel/libjxl/lib/jxl/dec_group.cc:413
#1  0x00007ff7dd6cf79f in jxl::DecodeGroup (readers=0x61f4ffc750,
    num_passes=1, group_idx=0, dec_state=0x61f4ffd3a0,
    group_dec_cache=0x1be61909680, thread=0, decoded=0x1be61905a20,
    first_pass=0, force_draw=false, dc_only=false)
    at C:/msys64/home/daniel/libjxl/lib/jxl/dec_group.cc:742
#2  0x00007ff7dd6b5f80 in jxl::FrameDecoder::ProcessACGroup (
    this=0x61f4ffcc50, ac_group_id=0, br=0x61f4ffc750, num_passes=1,
    thread=0, force_draw=false, dc_only=false)
    at C:/msys64/home/daniel/libjxl/lib/jxl/dec_frame.cc:570
#3  0x00007ff7dd6b6674 in operator() (__closure=0x61f4ffcaa0, g=0, thread=0)
    at C:/msys64/home/daniel/libjxl/lib/jxl/dec_frame.cc:738
#4  0x00007ff7dd6b90ce in jxl::ThreadPool::RunCallState<jxl::FrameDecoder::ProcessSections(const jxl::FrameDecoder::SectionInfo*, size_t, jxl::FrameDecoder::SectionStatus*)::<lambda(size_t)>, jxl::FrameDecoder::ProcessSections(const jxl::FrameDecoder::SectionInfo*, size_t, jxl::FrameDecoder::SectionStatus*)::<lambda(size_t, size_t)> >::CallDataFunc(void *, uint32_t, size_t) (
    jpegxl_opaque=0x61f4ffc900, value=0, thread_id=0)
    at C:/msys64/home/daniel/libjxl/lib/jxl/base/data_parallel.h:88
#5  0x00007ff7dd8daa96 in jpegxl::ThreadParallelRunner::Runner (
    runner_opaque=0x61f4fffa10, jpegxl_opaque=0x61f4ffc900,
    init=0x7ff7dd6b9058 <jxl::ThreadPool::RunCallState<jxl::FrameDecoder::ProcessSections(const jxl::FrameDecoder::SectionInfo*, size_t, jxl::FrameDecoder::SectionStatus*)::<lambda(size_t)>, jxl::FrameDecoder::ProcessSections(const jxl::FrameDecoder::SectionInfo*, size_t, jxl::FrameDecoder::SectionStatus*)::<lambda(size_t, size_t)> >::CallInitFunc(void *, size_t)>,
    func=0x7ff7dd6b9099 <jxl::ThreadPool::RunCallState<jxl::FrameDecoder::ProcessSections(const jxl::FrameDecoder::SectionInfo*, size_t, jxl::FrameDecoder::SectionStatus*)::<lambda(size_t)>, jxl::FrameDecoder::ProcessSections(const jxl::FrameDecoder::SectionInfo*, size_t, jxl::FrameDecoder::SectionStatus*)::<lambda(size_t, size_t)> >::CallDataFunc(void *, uint32_t, size_t)>, start_range=0,
    end_range=160)
    at C:/msys64/home/daniel/libjxl/lib/threads/thread_parallel_runner_internal.cc:72
#6  0x00007ff7dd6b8e78 in jxl::ThreadPool::Run<jxl::FrameDecoder::ProcessSections(const jxl::FrameDecoder::SectionInfo*, size_t, jxl::FrameDecoder::SectionStatus*)::<lambda(size_t)>, jxl::FrameDecoder::ProcessSections(const jxl::FrameDecoder::SectionInfo*, size_t, jxl::FrameDecoder::SectionStatus*)::<lambda(size_t, size_t)> >(uint32_t, uint32_t, const struct {...} &, const struct {...} &, const char *) (this=0x61f4fffa00, begin=0, end=160, init_func=...,
    data_func=..., caller=0x7ff7de248c0c <jxl::kTocDist+3836> "DecodeGroup")
    at C:/msys64/home/daniel/libjxl/lib/jxl/base/data_parallel.h:51
#7  0x00007ff7dd6b8cb1 in jxl::RunOnPool<jxl::FrameDecoder::ProcessSections(const jxl::FrameDecoder::SectionInfo*, size_t, jxl::FrameDecoder::SectionStatus*)::<lambda(size_t)>, jxl::FrameDecoder::ProcessSections(const jxl::FrameDecoder::SectionInfo*, size_t, jxl::FrameDecoder::SectionStatus*)::<lambda(size_t, size_t)> >(jxl::ThreadPool *, uint32_t, uint32_t, const struct {...} &, const struct {...} &, const char *) (pool=0x61f4fffa00, begin=0, end=160, init_func=...,
    data_func=..., caller=0x7ff7de248c0c <jxl::kTocDist+3836> "DecodeGroup")
    at C:/msys64/home/daniel/libjxl/lib/jxl/base/data_parallel.h:117
#8  0x00007ff7dd6b731f in jxl::FrameDecoder::ProcessSections (
    this=0x61f4ffcc50, sections=0x1be61860490, num=165,
    section_status=0x1be61862da0)
    at C:/msys64/home/daniel/libjxl/lib/jxl/dec_frame.cc:718
#9  0x00007ff7dd6b3558 in jxl::DecodeFrame (dparams=...,
    dec_state=0x61f4ffd3a0, pool=0x61f4fffa00, reader=0x61f4ffd360,
    decoded=0x1be61905a20, metadata=..., constraints=0x61f4ffec80,
    is_preview=false) at C:/msys64/home/daniel/libjxl/lib/jxl/dec_frame.cc:208
#10 0x00007ff7dd6879e0 in jxl::DecodeFile (dparams=..., file=...,
    io=0x61f4ffec80, pool=0x61f4fffa00)
    at C:/msys64/home/daniel/libjxl/lib/jxl/dec_file.cc:153
#11 0x00007ff7dd683479 in jpegxl::tools::DecompressJxlToPixels (
    compressed=..., params=..., pool=0x61f4fffa00, io=0x61f4ffec80,
    stats=0x61f4fff9d0) at C:/msys64/home/daniel/libjxl/tools/djxl.cc:189
#12 0x00007ff7dd681f30 in jpegxl::tools::(anonymous namespace)::DecompressMain
    (argc=4, argv=0x1be617c37f0)
    at C:/msys64/home/daniel/libjxl/tools/djxl_main.cc:178
#13 0x00007ff7dd6822b8 in main (argc=4, argv=0x1be617c37f0)
    at C:/msys64/home/daniel/libjxl/tools/djxl_main.cc:208
```

**Comment by novomesk** [comment]
It works for me now. Probably fixed by Highway upgrade to 0.15.0
